### ISSUE TITLE
Add a /flag <country name> for your country :)

### ISSUE THREAD

**Issue by PureKrome** [JabbR/JabbR#92]
Yo.

maybe have a tiny flag BEFORE the gravatar, on the left hand side.

also, autocomplete would be nice, for the country list ?

flags would be a single sprite .. and using css magical-unicorn-pewpew .. can display the correct one.

KThxBai !

/me hugs.

**Comment by davidfowl** [comment]
You gonna contribute this?

**Comment by PureKrome** [comment]
Hi David - sure :) Assign this bad boy to me.

What i was thinking was ..

/flag <[ISO 3166-1 alpha2](http://en.wikipedia.org/wiki/ISO_3166-1_alpha-2)>
/flag <no iso> removes any existing flag, if one was set.

ISO blah blah blah 2 == 2 char codes. US (for USA), AU (Australia), FR (France), etc...

For the ChatUser I was thinking of making this an ENUM (but that would a bit painful to make, the first time) or just a string. The DB will be an tinyint (i think they are under 255) or a smallint .. or char(2) if a string.

For the graphics, a [single image which will contain all the flags](http://flag-sprites.com/).  <-- ha! awesome .. already exists.

![flags](http://i.imgur.com/2T2VS.png)

Lastly, I suppose I was going to reference the gravatar / afk / note code to render the correct flag, if a flag has been set.

With Christmas here (in 1hr, 19 min...) i'll be afk for a few days as i'm wifed and kid'ed :~( but after that I can get straight onto it :)

**Comment by davidfowl** [comment]
Is that image ok to use? (licensing etc)

Merry Christmas!

**Comment by PureKrome** [comment]
I believe so. The license info is fond on the front page, right hand side .. towards the bottom..

> License
> FamFamFam flag icons are "available for free use for any purpose with no requirement for attribution"
> Blogpotato.de flag icons are licensed under Creative Commons
> Author and license terms for Maxmind icon set are unknown.

![](http://i.imgur.com/RjBFw.jpg)

**Comment by PureKrome** [comment]
David, I'm doing a check on the user's input to make sure that the IsoCode they have provided is legit.

To do this, i'm basically grabbing the first 2 chars of their optional text. Then I was going to do an almighty switch statement .. like. ..

switch (isoCode)
{
case "au":
case "ab":
...
return;
default: throw new InvalidOperationException("blah blah iso code is invalid blah blah");
}

.. now .. this switch statement will be long. As such, before I do this .. is there a better way I should handle this? Other ideas are to use an inmemory list and at some point, load that list .. linq against it .. but that sounds like an overkill.

Suggestions?

**Comment by davidfowl** [comment]
Only other way would be storing a hashset or dictionary.

**Comment by PureKrome** [comment]
Yep - that's what i was thinking when i was mentioning an in-memory list.

I'll do that .. because this means we can at least reuse that dictionary for other things, if we need (ie. list of available countries/iso's... etc).

thanks mate!

**Comment by PureKrome** [comment]
David, i'm just about to start coding in a `private StringDictionary` (http://msdn.microsoft.com/en-us/library/system.collections.specialized.stringdictionary.aspx) in the top of ChatServices.cs, fyi.

It could make the top of the file a bit ...long, So i'll wrap it within a `#region #endregion` to make the file more readable.

Alternatively, I can create a new static class (or a partial class) which can contain just this StringDictionary, if you want.

**Comment by davidfowl** [comment]
Just use `Dictionary<string, string>` and never use regions in this project.

**Comment by PureKrome** [comment]
Done.

Now, I'd like to show the country name for the flag in the tooltip .. so if a person hovers over it .. you can see the country.
Currently, the massive `IDictionary` exists in the `ChatService.cs`. I also don't want to repeat myself (DRY) with copying this dictionary into the javascript. So .. I was wondering if you might be ok with my adding a property into the ChatUser which might be used to _pass_ the country value up to the UI.

eg.

`[NotMapped]
public string CountryName { get; set; }`

and i set this at some point.

thoughts?

**Comment by PureKrome** [comment]
I think it's looking ok now :) 

I have only one bug - when a ChatUser enters a room, their tooltip is not showing. Delving into the code, I can see why this is :-

![](http://i.imgur.com/hxIWK.png)

Notice how there's some missing strings in there? Eg. Country? 

What I can't figure out is how this javascript method is called. Looking at the stack trace ...

![](http://i.imgur.com/IhWHW.png)

it shows me the two javascript method that are called .... before I get lost.

Can you tell me where I should be looking to seeing where the User data is passed in .. before it gets converted to the JavaScript userViewModel, please?

![](http://i.imgur.com/ZGQHJ.png)

**Comment by PureKrome** [comment]
Hi David.

I've fixed those two PR comments (was really simple) .. but I noticed that there was a visual bug with the Flag and AFK/Note icons. They were overlaying each other.

As such, I happen to have a mate over for NYE at my place and he's a CSS Gandalf, so instead of me wasting a lot of time, we paired and fixed up some of the CSS in 20 or so mins.

here's a screen shot of an edge case:

![](http://i.imgur.com/Rnnbc.png)

So we've split all the user stuff into a new div so when we have a really log name, it now wraps nicely and keeps the gravatar fixed to the top of the line item / row.

BTW, that scroll bar to the right .. that's NOT a scroll bar inside the web page but my browser scroll bar.

Now here's a screen shot of a really really long name....

![](http://i.imgur.com/Mae2z.png)

and here's one with spaces...

![](http://i.imgur.com/O21LN.png)
